controllers/reaper/reaper_controller.go

```diff
@@ -0,0 +1,347 @@
+	if !utils.CompareAnnotations(actualDeployment, desiredDeployment, k8ssandraapi.ResourceHashAnnotation) {
+		logger.Info("Updating Reaper Deployment")
+		resourceVersion := actualDeployment.GetResourceVersion()
+		desiredDeployment.DeepCopyInto(actualDeployment)
+		actualDeployment.SetResourceVersion(resourceVersion)
+		if err := controllerutil.SetControllerReference(actualReaper, actualDeployment, r.Scheme); err != nil {
+			logger.Error(err, "Failed to set controller reference on updated Reaper Deployment")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+		} else if err := r.Update(ctx, actualDeployment); err != nil {
+			logger.Error(err, "Failed to update Reaper Deployment")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+		} else {
+			logger.Info("Reaper Deployment updated successfully")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+		}
+	}
+
+	logger.Info("Reaper Deployment ready")
+	return ctrl.Result{}, nil
+}
+
+func (r *ReaperReconciler) reconcileService(
+	ctx context.Context,
+	actualReaper *reaperapi.Reaper,
+	logger logr.Logger,
+) (ctrl.Result, error) {
+	serviceKey := types.NamespacedName{Namespace: actualReaper.Namespace, Name: reaper.GetServiceName(actualReaper.Name)}
+	logger = logger.WithValues("Service", serviceKey)
+	logger.Info("Reconciling Reaper Service")
+	desiredService := reaper.NewService(serviceKey, actualReaper)
+	actualService := &corev1.Service{}
+	if err := r.Client.Get(ctx, serviceKey, actualService); err != nil {
+		if errors.IsNotFound(err) {
+			// create the service
+			if err = controllerutil.SetControllerReference(actualReaper, desiredService, r.Scheme); err != nil {
+				logger.Error(err, "Failed to set controller reference on Reaper Service")
+				return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+			}
+			logger.Info("Creating Reaper service")
+			if err = r.Client.Create(ctx, desiredService); err != nil {
+				if errors.IsAlreadyExists(err) {
+					// the read from the local cache didn't catch that the resource was created
+					// already; simply requeue until the cache is up-to-date
+					return ctrl.Result{Requeue: true}, nil
+				} else {
+					logger.Error(err, "Failed to create Reaper Service")
+					return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+				}
+			}
+			logger.Info("Reaper Service created successfully")
+			return ctrl.Result{}, nil
+		} else {
+			logger.Error(err, "Failed to get Reaper Service")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+		}
+	}
+	if !utils.CompareAnnotations(actualService, desiredService, k8ssandraapi.ResourceHashAnnotation) {
+		logger.Info("Updating Reaper Service")
+		updatedService := actualService.DeepCopy()
+		desiredService.DeepCopyInto(updatedService)
+		updatedService.SetResourceVersion(actualService.GetResourceVersion())
+		updatedService.Spec.ClusterIP = actualService.Spec.ClusterIP
+		updatedService.Spec.ClusterIPs = actualService.Spec.ClusterIPs
```

Comment:
This might be problematic. `ClusterIP` (and I think `ClusterIPs a well) is immutable.

Reply:
Well that's why I did this. If you don't keep the same values that were assigned upon service creation, the update will fail. I actually think that we need to do the same for other services, e.g. Stargate services.